python/lsst/meas/modelfit/optimizer/optimizer.cc

```diff
@@ -0,0 +1,207 @@
+using PyOptimizerControl = py::class_<OptimizerControl, std::shared_ptr<OptimizerControl>>;
+using PyOptimizerHistoryRecorder =
+    py::class_<OptimizerHistoryRecorder, std::shared_ptr<OptimizerHistoryRecorder>>;
+using PyOptimizer = py::class_<Optimizer, std::shared_ptr<Optimizer>>;
+
+PyOptimizerObjective declareOptimizerObjective(py::module & mod) {
+    PyOptimizerObjective cls(mod, "OptimizerObjective");
+    cls.def_readonly("dataSize", &OptimizerObjective::dataSize);
+    cls.def_readonly("parameterSize", &OptimizerObjective::parameterSize);
+    cls.def_static("makeFromLikelihood", &OptimizerObjective::makeFromLikelihood,
+                   "likelihood"_a, "prior"_a=py::none());
+    // class is abstract and not subclassable in Python, so we don't wrap the ctor
+    cls.def("fillObjectiveValueGrid", &OptimizerObjective::fillObjectiveValueGrid,
+            "parameters"_a, "output"_a);
+    cls.def("computeResiduals", &OptimizerObjective::computeResiduals,
+            "parameters"_a, "residuals"_a);
+    cls.def("differentiateResiduals", &OptimizerObjective::differentiateResiduals,
+            "parameters"_a, "derivatives"_a);
+    cls.def("hasPrior", &OptimizerObjective::hasPrior);
+    cls.def("computePrior", &OptimizerObjective::computePrior, "parameters"_a);
+    cls.def("differentiatePrior", &OptimizerObjective::differentiatePrior,
+            "parameters"_a, "gradient"_a, "hessian"_a);
+    return cls;
+}
+
+PyOptimizerControl declareOptimizerControl(py::module & mod) {
+    PyOptimizerControl cls(mod, "OptimizerControl");
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, noSR1Term);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, skipSR1UpdateThreshold);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, minTrustRadiusThreshold);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, gradientThreshold);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, numDiffRelStep);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, numDiffAbsStep);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, numDiffTrustRadiusStep);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, stepAcceptThreshold);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionInitialSize);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionGrowReductionRatio);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionGrowStepFraction);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionGrowFactor);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionShrinkReductionRatio);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionShrinkFactor);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionSolverTolerance);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, maxInnerIterations);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, maxOuterIterations);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, doSaveIterations);
```

Comment:
Happily most of these will probably never be changed - I just figured it was better to put them in the config class than define them as class- or module-level constants, just in case.  And the docs aren't great, but they do exist - they're just in the strings in the macros rather than comments.